### PR TITLE
Implement TOTP MFA

### PR DESCRIPTION
**Pull request by riamaria**
Rebased off of Kat's branch `zkat/uploads`

### CODE REVIEW

File: lib/banchan_web/live/auth_live/setup_mfa_live.ex [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  end
+
+  @impl true
+  def handle_event("setup_mfa", _, socket) do
+    user = socket.assigns.current_user
+    if user.totp_activated == true do
+      socket =
+        socket
+        |> put_flash(:info, "TOTP already activated")
+        |> redirect(to: Routes.setup_mfa_path(socket, :edit))
```
Comment (zkat) [inline lib/banchan_web/live/auth_live/setup_mfa_live.ex:83]: Are you sure you need to do all this redirecting? It's a liveview so it should auto-update, no?

File: lib/banchan_web/live/auth_live/setup_mfa_live.ex [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    socket = assign_defaults(session, socket)
+
+    user = socket.assigns.current_user
+
+    if user.totp_secret do
+      totp_uri = NimbleTOTP.otpauth_uri("Banchan:#{user.email}", user.totp_secret, issuer: "Banchan")
+      secret = Base.encode32(user.totp_secret, padding: false)
+      qr_code = QRCode.create!(totp_uri)
+      svg = qr_code |> QRCode.Svg.to_base64()
+      totp_svg = "data:image/svg+xml;base64,#{svg}"
```
Comment (zkat) [inline lib/banchan_web/live/auth_live/setup_mfa_live.ex:25]: magic!

File: lib/banchan_web/live/auth_live/setup_mfa_live.ex [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+          {@secret}
+        </div>
+        <div class="p-6">
+          <h1 class="text-2xl">Confirm MFA 6-digit OTP</h1>
+          <Form class="col-span-1" for={:user} submit="confirm_mfa">
+            <TextInput name={:token} label="One Time Password" opts={required: true} />
+            <Submit label="Activate" />
+          </Form>
+        </div>
+        {#elseif @qrcode_svg && @totp_activated}
```
Comment (zkat) [inline lib/banchan_web/live/auth_live/setup_mfa_live.ex:55]: I suspect you might need to run a `mix format`?

File: lib/banchan_web/live/live_helpers.ex [hunk 4]
```diff
@@ -17,7 +17,7 @@ defmodule BanchanWeb.LiveHelpers do
         find_current_user(session)
       end)
 
-    if auth && is_nil(socket.assigns.current_user) do
+    if auth && (is_nil(socket.assigns.current_user) || session.mfa_required) do
```
Comment (zkat) [inline lib/banchan_web/live/live_helpers.ex:20]: Can you explain this line to me?

**Review by zkat — COMMENTED** [review]
Please undelete the package-lock.json! We want that committed to the repo!

File: lib/banchan_web/live/live_helpers.ex (hunk 4 above)
Comment (riamaria) [reply lib/banchan_web/live/live_helpers.ex:20]: this was a mistake, i'm removing it now

File: lib/banchan_web/live/auth_live/setup_mfa_live.ex (hunk 1 above)
Comment (riamaria) [reply lib/banchan_web/live/auth_live/setup_mfa_live.ex:76]: In this case, we do need the redirecting. Otherwise the updating does not really happen

**Pull request closed without merging**